packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,375 @@
+import { ComponentType } from 'react';
+import {
+  ApplySchemaAttributes,
+  command,
+  CommandFunction,
+  extension,
+  ExtensionPriority,
+  ExtensionTag,
+  getTextSelection,
+  Handler,
+  isNumber,
+  NodeExtension,
+  NodeExtensionSpec,
+  NodeSpecOverride,
+  omitExtraAttributes,
+  PrimitiveSelection,
+  ProsemirrorNode,
+  ProsemirrorPlugin,
+  Transaction,
+} from '@remirror/core';
+import { NodeType } from '@remirror/pm/model';
+import { PasteRule } from '@remirror/pm/paste-rules';
+import { NodeViewComponentProps } from '@remirror/react';
+
+import { FileComponent, FileComponentProps } from './file-component';
+import { FileLoader } from './file-loader';
+import { defaultOptions } from './file-loaders/default';
+import {
+  findPlaceholderPlayload,
+  findPlaceholderPos,
+  placeholderPlugin,
+  setPlaceholderAction,
+} from './file-placeholder-plugin';
+
+export interface FileOptions {
+  loadFileHandler?: LoadFileHandler;
+  getAttrsBeforeUpload?: GetAttrsBeforeUpload;
+  getAttrsAfterUpload?: GetAttrsAfterUpload;
+  render?: (props: FileComponentProps) => React.ReactElement<HTMLElement> | null;
```

Comment:
Could render return a `HTMLElement` but not `React.ReactElement`?

Reply:
Users can use a React Component or a ProseMirror NodeView, because both of them have life cycle functions. But HTMLElement doesn't have life cycle functions. If users want to use NodeView, then just extend the FileExtension and implement the `createNodeView` method and remove `ReactComponent` should be simple enough.

Life cycle functions are needed because the component needs to register a callback function to the `FileLoader` so that the component can be updated when the upload progress is updated.